pulsar-broker/src/test/java/org/apache/pulsar/client/impl/InjectedClientCnxClientBuilder.java

```diff
@@ -0,0 +1,416 @@
+import org.apache.pulsar.common.api.proto.CommandActiveConsumerChange;
+import org.apache.pulsar.common.api.proto.CommandAddPartitionToTxnResponse;
+import org.apache.pulsar.common.api.proto.CommandAddSubscriptionToTxnResponse;
+import org.apache.pulsar.common.api.proto.CommandAuthChallenge;
+import org.apache.pulsar.common.api.proto.CommandCloseConsumer;
+import org.apache.pulsar.common.api.proto.CommandCloseProducer;
+import org.apache.pulsar.common.api.proto.CommandConnected;
+import org.apache.pulsar.common.api.proto.CommandEndTxnOnPartitionResponse;
+import org.apache.pulsar.common.api.proto.CommandEndTxnOnSubscriptionResponse;
+import org.apache.pulsar.common.api.proto.CommandEndTxnResponse;
+import org.apache.pulsar.common.api.proto.CommandError;
+import org.apache.pulsar.common.api.proto.CommandGetLastMessageIdResponse;
+import org.apache.pulsar.common.api.proto.CommandGetOrCreateSchemaResponse;
+import org.apache.pulsar.common.api.proto.CommandGetSchemaResponse;
+import org.apache.pulsar.common.api.proto.CommandGetTopicsOfNamespaceResponse;
+import org.apache.pulsar.common.api.proto.CommandLookupTopicResponse;
+import org.apache.pulsar.common.api.proto.CommandMessage;
+import org.apache.pulsar.common.api.proto.CommandNewTxnResponse;
+import org.apache.pulsar.common.api.proto.CommandPartitionedTopicMetadataResponse;
+import org.apache.pulsar.common.api.proto.CommandPing;
+import org.apache.pulsar.common.api.proto.CommandProducerSuccess;
+import org.apache.pulsar.common.api.proto.CommandReachedEndOfTopic;
+import org.apache.pulsar.common.api.proto.CommandSendError;
+import org.apache.pulsar.common.api.proto.CommandSendReceipt;
+import org.apache.pulsar.common.api.proto.CommandSuccess;
+import org.apache.pulsar.common.api.proto.CommandTcClientConnectResponse;
+import org.apache.pulsar.common.api.proto.CommandTopicMigrated;
+import org.apache.pulsar.common.api.proto.CommandWatchTopicListSuccess;
+import org.apache.pulsar.common.api.proto.CommandWatchTopicUpdate;
+import org.apache.pulsar.common.util.netty.EventLoopUtil;
+
+public class InjectedClientCnxClientBuilder {
+
+    public static PulsarClientImpl create(final ClientBuilderImpl clientBuilder, final ClientCnxCustomizer customizer)
+            throws Exception {
+        ClientConfigurationData conf = clientBuilder.getClientConfigurationData();
+        ThreadFactory threadFactory = new ExecutorProvider
+                .ExtendedThreadFactory("pulsar-client-io", Thread.currentThread().isDaemon());
+        EventLoopGroup eventLoopGroup =
+                EventLoopUtil.newEventLoopGroup(conf.getNumIoThreads(), conf.isEnableBusyWait(), threadFactory);
+
+        // Inject into ClientCnx.
+        ConnectionPool pool = new ConnectionPool(conf, eventLoopGroup,
+                () -> new InjectedClientCnx(conf, eventLoopGroup, customizer));
+
+        return new PulsarClientImpl(conf, eventLoopGroup, pool);
+    }
+
+    public static abstract class ClientCnxCustomizer {
```

Comment:
I don't see a reason to have ClientCnxCustomizer and InjectedClientCnx. Simply providing a `BiFunction<ClientConfigurationData, EventLoopGroup, ClientCnx>` parameter for creating the ClientCnx instance should be sufficient.
For example, you'd use it somewhat like this:
```
.clientCnxFunction((conf, eventLoopGroup) -> {
    return new ClientCnx(conf, eventLoopGroup) {
         @Override
        protected void handlePing(CommandPing ping) {
            // do not respond to CommandPing
        }
       };
});
```
The PulsarTestClient shows how it could be implemented. PulsarTestClient is specialized for a certain test, but it contains some useful ideas how clientCnxFunction could be used to instantiate the ClientCnx when it's needed.